codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriterTest.java

```diff
@@ -0,0 +1,282 @@
+        frameWriter.writeHeaders(ctx, streamId, headers, 0, true, promise);
+
+        byte[] expectedPayload = headerPayload(streamId, headers);
+
+        // First frame: HEADER(length=0x4000, flags=0x01)
+        assertEquals(Http2CodecUtil.MAX_FRAME_SIZE_LOWER_BOUND,
+                     outbound.readUnsignedMedium());
+        assertEquals(0x01, outbound.readByte());
+        assertEquals(0x01, outbound.readByte());
+        assertEquals(streamId, outbound.readInt());
+
+        byte[] firstPayload = new byte[Http2CodecUtil.MAX_FRAME_SIZE_LOWER_BOUND];
+        outbound.readBytes(firstPayload);
+
+        int remainPayloadLength = expectedPayload.length - Http2CodecUtil.MAX_FRAME_SIZE_LOWER_BOUND;
+        // Second frame: CONTINUATION(length=remainPayloadLength, flags=0x04)
+        assertEquals(remainPayloadLength, outbound.readUnsignedMedium());
+        assertEquals(0x09, outbound.readByte());
+        assertEquals(0x04, outbound.readByte());
+        assertEquals(streamId, outbound.readInt());
+
+        byte[] secondPayload = new byte[remainPayloadLength];
+        outbound.readBytes(secondPayload);
+
+        assertArrayEquals(Arrays.copyOfRange(expectedPayload, 0, firstPayload.length),
+                          firstPayload);
+        assertArrayEquals(Arrays.copyOfRange(expectedPayload, firstPayload.length,
+                                             expectedPayload.length),
+                          secondPayload);
+    }
+
+    @Test
+    public void writeFrameZeroPayload() throws Exception {
+        frameWriter.writeFrame(ctx, (byte) 0xf, 0, new Http2Flags(), Unpooled.EMPTY_BUFFER, promise);
+
+        byte[] expectedFrameBytes = {
+                (byte) 0x00, (byte) 0x00, (byte) 0x00, // payload length
+                (byte) 0x0f, // payload type
+                (byte) 0x00, // flags
+                (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00 // stream id
+        };
+
+        expectedOutbound = Unpooled.wrappedBuffer(expectedFrameBytes);
+        assertEquals(expectedOutbound, outbound);
+    }
+
+    @Test
+    public void writeFrameHasPayload() throws Exception {
+        byte[] payload = {(byte) 0x01, (byte) 0x03, (byte) 0x05, (byte) 0x07, (byte) 0x09};
+        ByteBuf payloadByteBuf = Unpooled.wrappedBuffer(payload);
+
+        try {
+            frameWriter.writeFrame(ctx, (byte) 0xf, 0, new Http2Flags(), payloadByteBuf, promise);
+
+            byte[] expectedFrameHeaderBytes = {
+                    (byte) 0x00, (byte) 0x00, (byte) 0x05, // payload length
+                    (byte) 0x0f, // payload type
+                    (byte) 0x00, // flags
+                    (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00 // stream id
+            };
+            expectedOutbound = Unpooled.copiedBuffer(expectedFrameHeaderBytes, payload);
+            assertEquals(expectedOutbound, outbound);
+        } finally {
+            ReferenceCountUtil.safeRelease(payloadByteBuf);
```

Comment:
we are relying upon the release behavior of the `frameWriter` in these tests so this release operation is expected to fail during normal operation. can you remove this?